### ISSUE TITLE
Confusing behaviour of `BufferDict.add_distribution`

### ISSUE THREAD

**Issue by Gattocrucco** [gplepage/gvar#20]
Consider the following

```python
import gvar
gvar.BufferDict.add_distribution('cippa', gvar.exp)
a = gvar.BufferDict({'cippa(x)': 1})
print(a['x'])
gvar.BufferDict.add_distribution('cippa', gvar.log)
print(a['x'])
a.buf    # has a side effect!
print(a['x'])
```

 Output:

```
2.718281828459045
2.718281828459045
0.0
```

Two problems:

  * `gvar.add_distribution` lets me silently redefine the distribution without an error, unlike `gvar.BufferDict.uniform`. This implicitly changes the values stored in existing `BufferDict`s.
  * However, the `BufferDict` instance "notices" the change only after I access its `buf` attribute. What's happening?

**Comment by gplepage** [comment]
This is a bug as well and will be fixed shortly. The problem is that ``a`` caches a value for ``a['x']`` when it is first accessed, so it needn't be reevaluated if accessed repeatedly. I didn't anticipate someone redefining an existing function after using the original definition , thereby invalidating the cached value. This is easily fixed. I will also add a warning if someone redefines a distribution without first deleting the old definition (which will work when I fix it).

**Comment by Gattocrucco** [comment]
Yes, it's a peculiar use case. It happened to me with very long scripts doing many fits.

**Comment by Gattocrucco** [comment]
Another thing: maybe for coherency with `BufferDict.uniform` you should raise an error instead of just a warning. Since these code paths are buggy, I think that breaking backward compatibility may not be a serious concern. Raising an error is the safer behaviour.

**Comment by gplepage** [comment]
These issues are addressed by version 11.9.6 which I just posted. Thanks for pointing them out.
